src/phongo_apm.c

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "bson/bson.h"
+#include "mongoc/mongoc.h"
+
+#include <php.h>
+#include <Zend/zend_hash.h>
+#include <Zend/zend_exceptions.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+#include "phongo_apm.h"
+
+#undef MONGOC_LOG_DOMAIN
+#define MONGOC_LOG_DOMAIN "PHONGO"
+
+ZEND_EXTERN_MODULE_GLOBALS(mongodb)
+
+/* Ensures that all subscribers in @from that are instances of @subscriber_ce
+ * are added to the set of subscribers in @to. This is used to build the list of
+ * subscribers to notify for an event. */
+static void phongo_apm_add_subscribers(zend_class_entry* subscriber_ce, HashTable* from, HashTable* to)
+{
+	zval* subscriber;
+
+	ZEND_HASH_FOREACH_VAL_IND(from, subscriber)
+	{
+		if (Z_TYPE_P(subscriber) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(subscriber), subscriber_ce)) {
+			continue;
+		}
+
+		if (zend_hash_index_exists(to, Z_OBJ_HANDLE_P(subscriber))) {
+			continue;
+		}
+
+		zend_hash_index_update(to, Z_OBJ_HANDLE_P(subscriber), subscriber);
+		Z_ADDREF_P(subscriber);
```

Comment:
I originally wasn't sure if it was absolutely necessary to add ref-counts while building this HashTable, but I think it's reasonable given a possible edge case where a subscriber could otherwise get garbage collected while dispatching a single event. Consider:

 * Register a global subscriber, such that the only reference to it is in the internal, global HashTable
 * Register a Manager subscriber
 * Execute a command on the Manager, which dispatches an event
 * The Manager subscriber gets notified first and unregisters the global subscriber, which immediately frees the first subscriber
 * Segfault while attempting to notify the global subscriber

Granted, `phongo_apm_get_subscribers` collects global subscribers before Manager subscribers so I'd expect the global subscribers will always be notified first. If the edge case above is not possible as written, I assume we could just swap the example for two different global subscribers.